src/main/java/com/conveyal/r5/analyst/Grid.java

```diff
@@ -126,96 +126,154 @@ public Grid (int zoom, int width, int height, int north, int west) {
+
+            for (int worldx = lonToPixel(env.getMinX(), zoom); worldx <= lonToPixel(env.getMaxX(), zoom); worldx++) {
+
                 int x = worldx - west;
                 int y = worldy - north;
 
                 if (x < 0 || x >= width || y < 0 || y >= height) continue; // off the grid
 
                 Geometry pixel = getPixelGeometry(x + west, y + north, zoom);
-                Geometry intersection = pixel.intersection(geometry);
-                double denominator = relativeToPixels ? pixel.getArea() : area;
-                double weight = intersection.getArea() / denominator;
-                weights.put(new int[] { x, y }, weight);
+
+                if (pixelAreaAtLat == -1) pixelAreaAtLat = pixel.getArea(); //Recalculate for a new latitude.
+
+                if (preparedGeom.intersects(pixel)){ // pixel is at least partly inside the feature
+                    Geometry intersection = pixel.intersection(geometry);
+                    double denominator = relativeToPixels ? pixelAreaAtLat : area;
+                    double weight = intersection.getArea() / denominator;
+                    if (pixelWeightCallback == null){
+                        weights.add(new PixelWeight(x, y, weight));
+                    } else {
+                        pixelWeightCallback.handlePixelWeight(x,y,weight);
+                    }
+                }
             }
         }
-
         return weights;
     }
 
     /**
      * Do pycnoplactic mapping:
-     * the value associated with the supplied polygon a polygon will be split out proportionately to
+     * the value associated with the supplied polygon will be split out proportionately to
      * all the web Mercator pixels that intersect it.
      *
      * If you are creating multiple grids of the same size for different attributes of the same input features, you should
      * call getPixelWeights(geometry) once for each geometry on any one of the grids, and then pass the returned weights
      * and the attribute value into incrementFromPixelWeights function; this will avoid duplicating expensive geometric
      * math.
      */
-    public void rasterize (Geometry geometry, double value) {
+    private void rasterize (Geometry geometry, double value) {
         incrementFromPixelWeights(getPixelWeights(geometry), value);
     }
 
     /** Using a grid of weights produced by getPixelWeights, burn the value of a polygon into the grid */
-    public void incrementFromPixelWeights (TObjectDoubleMap<int[]> weights, double value) {
-        for (TObjectDoubleIterator<int[]> it = weights.iterator(); it.hasNext();) {
-            it.advance();
-            grid[it.key()[0]][it.key()[1]] += it.value() * value;
+    public void incrementFromPixelWeights (ArrayList weights, double value) {
+        Iterator<PixelWeight> pixelWeightIterator = weights.iterator();
+        while(pixelWeightIterator.hasNext()) {
```

Comment:
You can use the foreach syntax here: `for (PixelWeight pix : weights) { ... }`. You only need to create an iterator explicitly when you want to call methods on that iterator, e.g. removing elements.